cocotb/types/range.py

```diff
@@ -0,0 +1,175 @@
+from typing import Any, Iterator, overload
+from collections.abc import Sequence
+from functools import cached_property, lru_cache
+
+
+class Range(Sequence):
+    r"""
+    Variant of :class:`range` with inclusive right bound.
+
+    In Python, :class:`range` and :class:`slice` have a non-inclusive right bound.
+    In both Verilog and VHDL, ranges and arrays have an inclusive right bound.
+    This type mimic's Python's :class:`range` type, but supports HDL-like inclusive right bounds.
+    Also supports :attr:`left`, :attr:`right`, and :attr:`length` attributes as seen in VHDL.
+
+    .. code-block:: python3
+
+        >>> r = Range(-2, 3)
+        >>> r.left, r.right, r.length
+        (-2, 3, 6)
+
+        >>> s = Range(8, 'downto', 1)
+        >>> s.left, s.right, s.length
+        (8, 1, 8)
+
+    :meth:`from_range` and :meth:`to_range` can be used to convert from and to :class:`range`.
+
+    .. code-block:: python3
+
+        >>> r = Range(-2, 3)
+        >>> r.to_range()
+        range(-2, 4)
+
+    Supports "null" ranges as seen in VHDL.
+    "null" ranges occur when a left bound cannot reach a right bound with the given direction.
+    They have a length of 0, but the left, right, and direction values remain as given.
+
+    .. code-block:: python3
+
+        >>> r = Range(1, 'to', 0)  # no way to count from 1 'to' 0
```

Comment:
What are your thoughts on the spelling `Range(1).to(0)`, where `type(Range).__call__` is overloaded to return some dummy object?

Reply:
It's neat, but it makes it hard to use an existing range to feed into another range if we do that.

```python
getattr(Range(r.left - 3), r.direction)(r.right)
# vs
Range(r.left - 3, r.direction, r.right)
```

Reply:
That's a fair argument.

I suppose another gimmick you could play, but probably not in this PR, is `R[3:"downto":1]` as a shorthand, but the string literal makes me uneasy for the same reason.

Reply:
Perhaps `Range[3:1]` `Range[3:1:-1]` syntax can be added as a follow on.